src/Xrefcheck/System.hs

```diff
@@ -51,125 +45,133 @@ askWithinCI = lookupEnv "CI" <&> \case
+-- considered as part of the filename instead of denoting a path
+-- separator.
+newtype RelPosixLink = RelPosixLink
+  { unRelPosixLink :: Text
+  } deriving newtype (Show, Eq, Ord, NFData, Buildable, Pretty)
+
+-- | Create a POSIX file path from a platform-dependent one.
+mkRelPosixLink :: FilePath -> RelPosixLink
+mkRelPosixLink = RelPosixLink
+  . withPathSeparator FPP.pathSeparator
+  . fromString
+
+-- | Join two 'CanonicalRelPosixLink's.
+(</>) :: RelPosixLink -> RelPosixLink -> RelPosixLink
+RelPosixLink a </> RelPosixLink b =
+  let a' = fromMaybe a $ T.stripSuffix "/" a
+      b' = fromMaybe b $ T.stripPrefix "./" a
+  in case (a', b') of
+        ("", _) -> RelPosixLink b
+        (".", _) -> RelPosixLink b
+        _ -> RelPosixLink $ a' <> "/" <> b'
+
+-- Get the platform-dependent file path from a 'RelPosixLink'
+-- considered as relative to another given platform-dependent
+-- 'FilePath'.
 --
--- It returns Nothing if child cannot be reached from root downwards
--- in the filesystem tree.
+-- In Windows, every `\` occurrence will be replaced by `/`.
+filePathFromRoot :: FilePath -> RelPosixLink -> FilePath
+filePathFromRoot rootPath = (rootPath FP.</>)
+  . toString
+  . withPathSeparator FP.pathSeparator
+  . unRelPosixLink
+
+-- | 'FilePath.takeDirectory' version for 'RelPosixLink'.
+takeDirectory :: RelPosixLink -> RelPosixLink
+takeDirectory = RelPosixLink
+  . fromString
+  . FPP.takeDirectory
+  . toString
+  . unRelPosixLink
+
+-- | 'FilePath.takeExtension' version for 'RelPosixLink'.
+takeExtension :: RelPosixLink -> String
+takeExtension = FPP.takeExtension
+  . toString
+  . unRelPosixLink
+
+-- | 'Check if a 'RelPosixLink' contains any backslash.
+hasBackslash :: RelPosixLink -> Bool
+hasBackslash = ('\\' `elem`)
+  . unRelPosixLink
+
+-- | Get the list of directories between a 'RelPosixLink' and its
+-- relative root.
+getIntermediateDirs :: RelPosixLink -> [RelPosixLink]
+getIntermediateDirs link = fmap RelPosixLink $
+  case T.splitOn "/" $ unRelPosixLink $ takeDirectory link of
+    [] -> []
+    ["."] -> [""]
+    [".."] -> ["", ".."]
+    d : ds -> scanl (\a b -> a <> "/" <> b) d ds
+
+-- | Relative POSIX file path with some normalizations applied.
 --
--- The resulting `FilePath` uses POSIX path separators.
-getPosixRelativeChild :: CanonicalPath -> CanonicalPath -> Maybe FilePath
-getPosixRelativeChild (UnsafeCanonicalPath root) (UnsafeCanonicalPath child) =
-  dropLeadingSepAndEmptyCase . fmap replaceSeparator <$> stripPrefix root child
-  where
-    replaceSeparator :: Char -> Char
-    replaceSeparator c
-      | FP.isPathSeparator c = FPP.pathSeparator
-      | otherwise = c
-
-    dropLeadingSepAndEmptyCase :: FilePath -> FilePath
-    dropLeadingSepAndEmptyCase path = case dropWhile FP.isPathSeparator path of
-      "" -> "."
-      other -> other
+-- It should be created from a 'RelPosixLink' via
+-- 'canonicalizeRelPosixLink'.
+newtype CanonicalRelPosixLink = UnsafeCanonicalRelPosixLink
```

Comment:
Btw on the naming, aren't "canonical" and "posix" mutually constradicting properties? One is about system-dependent path-separators, and the other is about posix path-separators; or it's more than that?

Reply:
Hmm currently a canonical link is a valid relative POSIX link (i.e. `CanonicalRelPosixLink` subset of `RelPosixLink`). 

Maybe the name is an implementation detail leak, is that what you mean? I can change it to just `CanonicalLink` if it sounds more appropriate.

Reply:
Oh sorry, I see I was saying nonsense, please revert that commit.